auth/env_keychain_test.go

```diff
@@ -46,97 +46,151 @@ func testEnvKeychain(t *testing.T, when spec.G, it spec.S) {
 
-						h.AssertEq(t, header, &authn.AuthConfig{RegistryToken: "qwerty="})
-					})
+					_, err = envKeyChain.Resolve(registry)
+					h.AssertError(t, err, "failed to parse CNB_REGISTRY_AUTH value")
+				})
+			})
 
-					it("returns an Anonymous authenticator when the environment does not have a auth header", func() {
-						envKeyChain = auth.NewKeychain("CNB_REGISTRY_AUTH")
-						registry, err := name.NewRegistry("no-env-auth-registry.com", name.WeakValidation)
-						h.AssertNil(t, err)
+			when("env var is not set", func() {
+				it("returns an Anonymous authenticator", func() {
+					registry, err := name.NewRegistry("no-env-auth-registry.com", name.WeakValidation)
+					h.AssertNil(t, err)
 
-						authenticator, err := envKeyChain.Resolve(registry)
-						h.AssertNil(t, err)
+					auth, err := envKeyChain.Resolve(registry)
+					h.AssertNil(t, err)
 
-						h.AssertEq(t, authenticator, authn.Anonymous)
-					})
+					h.AssertEq(t, auth, authn.Anonymous)
 				})
+			})
+		})
+	})
 
-				when("invalid env var is set", func() {
-					it.Before(func() {
-						err := os.Setenv("CNB_REGISTRY_AUTH", "NOT -- JSON")
-						h.AssertNil(t, err)
-					})
+	when("#FallbackKeychain", func() {
+		when("#Resolve", func() {
+			when("DefaultKeychain resolves successfully", func() {
+				it("returns a non-anonymous authenticator", func() {
+					defaultKeychain := &FakeKeychain{
+						authMap: map[string]*authn.AuthConfig{
+							"basic-registry.com": {
+								Username: "user",
+								Password: "password",
+							},
+						},
+					}
+					fallbackKeychain := &auth.FallbackKeychain{DefaultKeychain: defaultKeychain}
+					registry, err := name.NewRegistry("basic-registry.com", name.WeakValidation)
+					h.AssertNil(t, err)
 
-					it("returns an error", func() {
-						registry, err := name.NewRegistry("some-registry.com", name.WeakValidation)
-						h.AssertNil(t, err)
+					authenticator, err := fallbackKeychain.Resolve(registry)
+					h.AssertNil(t, err)
 
-						_, err = envKeyChain.Resolve(registry)
-						h.AssertError(t, err, "failed to parse CNB_REGISTRY_AUTH value")
-					})
+					header, err := authenticator.Authorization()
+					h.AssertNil(t, err)
+					h.AssertEq(t, header.Username, "user")
+					h.AssertEq(t, header.Password, "password")
 				})
+			})
 
-				when("env var is not set", func() {
-					it("returns an Anonymous authenticator", func() {
-						registry, err := name.NewRegistry("no-env-auth-registry.com", name.WeakValidation)
-						h.AssertNil(t, err)
+			when("DefaultKeychain resolves with error", func() {
+				var dockerConfigEnv string
+				it.Before(func() {
+					dockerConfigEnv = os.Getenv("DOCKER_CONFIG")
+					// authn.DefaultKeychain will error if DOCKER_CONFIG directory is not readable
+					h.AssertNil(t, os.Setenv("DOCKER_CONFIG", "does-not-exist"))
+				})
+				it.After(func() {
+					h.AssertNil(t, os.Setenv("DOCKER_CONFIG", dockerConfigEnv))
+				})
+				it("returns an anonymous authenticator", func() {
```

Comment:
This test doesn't actually require `auth.FallbackKeychain` to succeed. `authn.DefaultKeychain` already return `authn.Anonymous` rather than an error when the docker config directory does not exist. Rather, I believe this issue occurs when we cannot determine whether or not the dir exists because it is located in `/home/root` and we do not have permissions to read `/home/root` after dropping privileges.

https://github.com/google/go-containerregistry/blob/4e3ad4635a12aa559961ef1c8cab4af8054d04b6/vendor/github.com/docker/cli/cli/config/config.go#L102-L106

Reply:
Our use of `DOCKER_CONFIG` here is a bit lazy. We were not concerned with how `DefaultKeychain.Resolve()` errored, only that an error occurred. We considered updating `FakeKeychain.Resolve()` to return an error on demand, which also would've satisfied this test. I think what this is exposing is that our rescuing from error cases is too broad. This should go away once we address your [other comment](https://github.com/buildpacks/lifecycle/pull/405#pullrequestreview-475844005).

Reply:
Oh, and now that I'm reading your comment again, I see this test is not actually testing what it is supposed to...